search-app/src/cmr/search/services/association_validation.clj

```diff
@@ -0,0 +1,260 @@
+        (append-error association (assoc-msg/inaccessible-collection-revision coll-revision))
+        association))))
+
+(defmulti validate-association-conflict-for-collection
+  "Validate the given tag association does not conflict with existing tag associations in that
+  a tag cannot be associated with a collection revision and the same collection without revision
+  at the same time."
+  (fn [context assoc-type key association]
+    assoc-type))
+
+(defmethod validate-association-conflict-for-collection :tag
+  [context assoc-type tag-key tag-association]
+  (let [tas (->> (mdb/find-concepts context
+                                    {:tag-key tag-key
+                                     :associated-concept-id (:concept-id tag-association)
+                                     :exclude-metadata true
+                                     :latest true}
+                                    :tag-association)
+                 (filter #(not (:deleted %))))
+        coll-revision-ids (map #(get-in % [:extra-fields :associated-revision-id]) tas)
+        not-nil-revision-ids (remove nil? coll-revision-ids)]
+    (cond
+      ;; there is no existing tag association found, no need to validate
+      (= 0 (count coll-revision-ids))
+      nil
+
+      ;; there are existing tag associations and they are all on collection revisions
+      (= (count coll-revision-ids) (count not-nil-revision-ids))
+      (when-not (:revision-id tag-association)
+        (format (str "There are already tag associations with tag key [%s] on collection [%s] "
+                     "revision ids [%s], cannot create tag association on the same collection without revision id.")
+                tag-key (:concept-id tag-association) (string/join ", " coll-revision-ids)))
+
+      ;; there are existing tag associations and they are all on collection without revision
+      (= 0 (count not-nil-revision-ids))
+      (when-let [revision-id (:revision-id tag-association)]
+        (format (str "There are already tag associations with tag key [%s] on collection [%s] "
+                     "without revision id, cannot create tag association on the same collection "
+                     "with revision id [%s].")
+                tag-key (:concept-id tag-association) revision-id))
+
+      ;; there are conflicts within the existing tag associations in metadata-db already
+      :else
+      (format (str "Tag can only be associated with a collection or a collection revision, "
+                   "never both at the same time. There are already conflicting tag associations "
+                   "in metadata-db with tag key [%s] on collection [%s] , please delete "
+                   "one of the conflicting tag associations.")
+              tag-key (:concept-id tag-association)))))
+
+(defmethod validate-association-conflict-for-collection :variable
+  [context assoc-type variable-name variable-association]
+  (let [tas (->> (mdb/find-concepts context
```

Comment:
I have tried to refactor duplicated code into common functions, and these functions are left alone as they are more readable this way. I will rename the `tas`s.